src/status_im/subs/wallet/screens/input_amount.cljs

```diff
@@ -0,0 +1,360 @@
+       upper-limit-equals-input-value?
+       no-routes-found?
+       limit-exceeded?
+       enabled-from-chain-ids]]
+   (let [owned-eth-token (rf/sub [:wallet/token-by-symbol
+                                  (string/upper-case constants/mainnet-short-name)
+                                  enabled-from-chain-ids])]
+     (insufficient-asset-amount?
+      {:enough-assets?                  enough-assets?
+       :token-symbol                    token-symbol
+       :owned-eth-token                 owned-eth-token
+       :upper-limit-equals-input-value? upper-limit-equals-input-value?
+       :no-routes-found?                no-routes-found?
+       :limit-exceeded?                 limit-exceeded?
+       :sender-network-values           sender-network-values}))))
+
+(rf/reg-sub :send-input-amount-screen/should-try-again?
+ :<- [:send-input-amount-screen/upper-limit-exceeded?]
+ :<- [:send-input-amount-screen/no-routes-found?]
+ :<- [:send-input-amount-screen/not-enough-asset?]
+ (fn [[limit-exceeded?
+       no-routes-found?
+       not-enough-asset?]]
+   (and (not limit-exceeded?)
+        no-routes-found?
+        (not not-enough-asset?))))
+
+(rf/reg-sub :send-input-amount-screen/show-no-routes?
+ :<- [:send-input-amount-screen/upper-limit-exceeded?]
+ :<- [:send-input-amount-screen/no-routes-found?]
+ :<- [:send-input-amount-screen/not-enough-asset?]
+ :<- [:wallet/wallet-send-sender-network-values]
+ (fn [[limit-exceeded?
+       no-routes-found?
+       not-enough-asset?
+       sender-network-values]]
+   (and (or no-routes-found? limit-exceeded?)
+        (not-empty sender-network-values)
+        (not not-enough-asset?))))
+
+(rf/reg-sub :send-input-amount-screen/confirm-disabled?
+ :<- [:wallet/wallet-send-route]
+ :<- [:send-input-amount-screen/valid-input?]
+ (fn [[route
+       valid-input?]]
+   (or (nil? route)
+       (empty? route)
+       (not valid-input?))))
+
+(defn- get-fee-formatted
+  [route]
+  (when-let [native-currency-symbol (-> route first :from :native-currency-symbol)]
+    (rf/sub [:wallet/wallet-send-fee-fiat-formatted native-currency-symbol])))
+
+(rf/reg-sub :send-input-amount-screen/fee-formatted
+ :<- [:wallet/wallet-send-route]
+ :<- [:send-input-amount-screen/confirm-disabled?]
+ :<- [:send-input-amount-screen/not-enough-asset?]
+ (fn [[route
+       confirm-disabled?
+       not-enough-asset?]]
+   (when (or (not confirm-disabled?) not-enough-asset?)
+     (get-fee-formatted route))))
+
+(rf/reg-sub :send-input-amount-screen/show-estimated-fees?
+ :<- [:wallet/wallet-send-route]
+ :<- [:send-input-amount-screen/no-routes-found?]
+ :<- [:send-input-amount-screen/not-enough-asset?]
+ :<- [:wallet/wallet-send-loading-suggested-routes?]
+ (fn [[route
+       no-routes-found?
+       not-enough-asset?
+       loading-routes?]]
+   (or (and (not no-routes-found?) (or loading-routes? route))
+       not-enough-asset?)))
+
+(rf/reg-sub :send-input-amount-screen/token-not-available?
```

Comment:
I’m not sure either, it looks a bit odd to me. Maybe a combination of general subscriptions and utility functions used in the UI layer could make the code more consistent with the rest of the codebase and improve readability.

Reply:
> subscription in most cases should be general and could be used in any part of the app

@flexsurfer, would you mind providing some reasoning for this statement? Because I tend to disagree here, the main purpose of re-frame subscriptions is to provide a way for a view to re-render when data changes. So when you have a data for specific screen, you can have a screen-specific subscription.

But of course I agree that in this file not all subscriptions should be tied to particular screen, some are related to the whole send flow.

Reply:
I'm also weary of this approach because it's a departure from what we've been doing for years and which has worked well, that is, subscriptions provide some computation on the state, but they don't care about which screen they are being used.

Going on the direction of @briansztamfater's feedback, _subscriptions for views_ are a new concept. These subs will change much more often because they will be more tied to screens, contrary to our current subs, which tend to be quite stable because they mostly need to change if the global state changes.

We could have a local file named `subs.cljs` in the same directory of the `view.cljs` file. Our agreement/convention would then be that the subscriptions in this localized subs file would not be reused by other views or other view subs, otherwise I believe things would get messy quickly. I'm not a fan of this approach, but I think it's better than mixing specific screens concerns with the general subs in `status-im.subs`.

My main concern is related to the complexity of the subs graph, which would explode in the future with lots of layer-3 subs depending on other layer-3 subs. The problem is that some subs will "pull the world" and small changes in subs will cause rippling effects in unwanted places. This has been a problem with large subs pulling from too many inputs and from other layer-3 subs, it's certainly going to happen with this approach, although nobody can predict by which extent. Maybe it will be manageable, but I'm still weary @vkjr

Reply:
> My main concern is related to the complexity of the subs graph, which would explode in the future with lots of layer-3 subs depending on other layer-3 subs. 

> The problem is that some subs will "pull the world" and small changes in subs will cause rippling effects in unwanted places.

@ilmotta, all the screen-specific subs that I've created were extracted from huge `let` bindings inside the view. Before extracting they were local computations that still dependent on non-screen-specific subs and on each other. So the rippling effect of changes in high-level subs is still there. Also all these computations were made **every time** when any value of the subs changed. By moving them to subscriptions in different views we reduce the amount of recomputations but rippling effect stays intact - it was here, it is still here. But maybe I'm getting you incorrect and you mean something else by "rippling effect"? Wdyt? 

>I'm also weary of this approach because it's a departure from what we've been doing for years and which has worked well

I would like to opine here - this approach also led us to 160-line `let` statement with tons of local computations :) 
For example we have a screen-specific state - value of input field. Also we have a screen-specific state - error flag if input value is higher than amount of tokens user has. Implementing this error flag with screen-specific subscription is trivial and moves logic out of the view. Implementing this without subscription require us to inflate view code with more logic.

But of course this depends on the approach we are comfortable with. I'd be happy to see the views more dumb and contain only basic logic related to displaying with all the other logic moved out)

Reply:
@vkjr my comment about pulling too much in subs (too many inputs, including other layer-3 subs) is about a general problem, and not so much about this PR (I haven't checked the runtime). We know from experience that subs that pull too much can be difficult to maintain in terms of performance. This is what happened to chat and community subs, but we moved eventually to more granular subs.

> Also all these computations were made every time when any value of the subs changed.

Large subs with too many inputs/dependencies will have a tendency to have a lot more cache misses. Not saying this is what happens with subs in this PR because I haven't checked.

As I suggested previously, this problem *could* be mitigated *sometimes* by having a clear standard for subs that should not be reused outside specific screens (via naming and/or file system structure). If these subs proposed in the PR are not reused across views then it might be fine because they will be discarded on unmount. Just so we avoid situations like this https://github.com/status-im/status-mobile/blob/76d6825dd568f11d3948d70290548cc50a312269/src/status_im/subs/communities.cljs#L21-L23

-------

> I would like to opine here - this approach also led us to 160-line let statement with tons of local computations :)

My 2 cents about `input-amount.view/view` is that it is one of the top offenders in the repo in terms of "quality". Not the only one, but it is an outlier. The thing to me is, there's nothing too complex about the screen using the input that justifies the style of code currently in `develop`. The approach we've used so far in other parts of the app over years has allowed us to build highly complex screens without the need to directly couple subs with screens (to some extent). It does look like we are treating the symptoms in our discussions, not the "disease".

I don't think the world will end if we follow what's proposed in this PR :) I just see we are skipping basic steps in code quality and the correct usage of existing recommendations for React, Reagent, re-frame, and Clojure and jumping straight into a pattern that will keep masking deeper problems. Yes, the views will be more dumb and that's a win, but it does look a bit like sweeping dirt under the rug.

One of the deeper problems to me is that the data models for the mobile wallet are not optimal; the endpoints in status-go also are not abstracting optimally for the mobile client (probably the same story for the desktop client as far as I've heard). We are talking a lot about the subs, but the events are suffering just as badly or worse in certain parts of the wallet. If we get the data models correctly and the right level of abstraction in endpoints, then we can remove tons of complex code in the client, (which we probably didn't test adequately).

--------

Another interesting analysis we can make is understand which parts of our process failed, that led us to produce what many of us deem unacceptable code. To give one example, there were over 70 commits that modified `src/status_im/contexts/wallet/send/input_amount/view.cljs` spread over almost 1 year. What can we do better to prevent so many problems to pile on top of each other for so long? Have PR reviews worked well in our opinion? If not, how can we improve?

-------

In summary @vkjr, I won't be convinced the proposal is good for us in the long run (as a general pattern, beyond this PR), but I trust the judgement from other CCs if more people think this proposal is good. I'm probably over-concerned.

Reply:
@ilmotta, I agree that we need to formulate some architecture goal toward which we are going to move as a team and engage more CCs to assess this goal 👍 

I think that significant part of the problem with this particular PR lays in the fact that I decided to keep very local UI state like input value inside the rf db which is probably overkill and I'm totally okay with giving up on this idea in favor of react hooks for example. 

And I agree with your take that `the re-frame layer should model data independently from the view layer` but to some extent. For example we have `send` flow that consists of few screens and we need to save the choices made on those screens - what token we are going to send, to what address and what amount. And I see no better choice than to store that in rf-db. But don't think we should store as much as we do now. For example we can save token symbol `:ETH` as a user choice and don't need to copy full token structure under `[:wallet :ui :send]` because there is another part of rf-db where that token can be found by symbol. Similarly we probably don't need to store `route` data on `ui` level of rf-db because that leads to UI containing logic on how to operate with that route structure while it can be done on lower levels.

So, if instead of squeezing the ugliness out of UI you would prefer to start with squeezing the ugliness out of data layer - I'm in :) I think that when we start to apply some rules at any level, other levels will also get better.

As I see it, we probably need to formulate requirements that can be applied to our data. And to formulate them we need to pose, discuss and answer bunch of questions. For example:
- what we want and do not want to store in rf-db?
- where local screen states should be stored?
- if we have flow that consists of few screens, where do we want to store the state of the flow?
- where navigation state should be stored?
- do we want the data received from status-go to be shaped before storing or we are storing it as-is and convert later?
- how we want to structure rf-db? plain? nested? What top-level keys or namespaces we can strictly assign to specific purposes?

wdyt?

Reply:
These are excellent questions to start with @vkjr.

Because there can be many answers to each question and even personal preferences, in my experience I saw that it helps if we also try to build/have guiding principles. We could call constraints, or even architecture depending on how we do. More than guidelines, they would play a role similar to the strategy being built for Status right now. Without that, answers can feel arbitrary or biased towards those that scream louder.

For example, if one of our constraints is that we should model the app-db more like a relational database doing schema on write, then that constraint by itself already guides us in answering a few questions:

- Less nested (not necessarily plain).
- Decode and transform when writing to the app-db. Consider malli schemas to represent some parts of the db (schema on write).
- Supporting normalization to facilitate efficient queries (subs) from nodes in the tree.
- Orthogonal concerns would be persisted (like navigation), but not local input state. Things in the middle could use a React Context. State traveling too many components could also use a React Context, but not necessarily re-frame (here the answer is probably "it depends" for a lot of cases).

I tend to like this theory of constraints or something of that kind to reduce the solution space and help establish coherent solutions. And we can have competing constraints that we can benchmark and challenge. Some might prefer a global state that behaves more like MongoDB, and so on.

It would be great if we could start with the problems and have a document to guide our progress towards these architectural discussions and eventually decisions. We could then resurrect `doc/decisions/` once we reach consensus. We might also want to have a small focus group of people that will drive this kind of initiative (somewhat like a crew (?), short-lived).

Reply:
@ilmotta, yeah, we definitely need to raise and discuss all these questions. Because my preference is to keep everything in rf-db including flow states and probably even ui states and doing this in a highly hierarchical way :) 
I see this approach more convenient because within reagent componets re-frame machinery like subscriptions look much nicer than react hooks. And it is much more convenient for developer to see all app state in rf-db.
So probably along with answering questions we will need also some experiments, like performances measurements to support point of view :)